src/target/riscv/riscv.c

```diff
@@ -1561,6 +1561,100 @@ int riscv_remove_watchpoint(struct target *target,
+		return M_HIT;
+	const riscv_reg_t tdata1_test = set_field(tdata1, CSR_MCONTROL_HIT, 1);
+	if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test) != ERROR_OK)
+		return M_HIT_ERROR;
+	riscv_reg_t tdata1_test_rb;
+	if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+		return M_HIT_ERROR;
+	int tdata1_test_hit = get_field(tdata1_test, CSR_MCONTROL_HIT);
+	int tdata1_test_rb_hit = get_field(tdata1_test_rb, CSR_MCONTROL_HIT);
+	if (tdata1_test_rb_hit != tdata1_test_hit)
+		return M_HIT_NOT_SUPPORTED;
+
+	if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
+		return M_HIT_ERROR;
+	return M_NOT_HIT;
+}
+
+typedef enum {
+	M6_HIT_ERROR,
+	M6_HIT_NOT_SUPPORTED,
+	M6_NOT_HIT,
+	M6_HIT_BEFORE,
+	M6_HIT_AFTER,
+	M6_HIT_IMM_AFTER
+} mctrl6hitstatus;
+
+static mctrl6hitstatus check_mcontrol6_hit_status(struct target *target,
+		riscv_reg_t tdata1, uint64_t hit_mask)
+{
+	int hit0 = get_field(tdata1, CSR_MCONTROL6_HIT0);
+	int hit1 = get_field(tdata1, CSR_MCONTROL6_HIT1);
+	int hit_info = (hit1 << 1) | hit0;
+	if (hit_info == CSR_MCONTROL6_HIT0_BEFORE)
+		return M6_HIT_BEFORE;
+
+	if (hit_info == CSR_MCONTROL6_HIT0_AFTER)
+		return M6_HIT_AFTER;
+
+	if (hit_info == CSR_MCONTROL6_HIT0_IMMEDIATELY_AFTER)
+		return M6_HIT_IMM_AFTER;
+
+	if (hit_info == CSR_MCONTROL6_HIT0_FALSE) {
+		/* Check timing "before" support. */
+		riscv_reg_t tdata1_hit_before_test = set_field(tdata1, CSR_MCONTROL6_HIT0, 1);
+		if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_hit_before_test) != ERROR_OK)
+			return M6_HIT_ERROR;
+		riscv_reg_t tdata1_test_rb;
+		if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+			return M6_HIT_ERROR;
+		if (tdata1_test_rb == tdata1_hit_before_test) {
+			if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
+				return M6_HIT_ERROR;
+			return M6_NOT_HIT;
+		}
+
+		/* Check timing "after" support. */
+		riscv_reg_t tdata1_hit_after_test = set_field(tdata1, CSR_MCONTROL6_HIT1, 1);
+		if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_hit_after_test) != ERROR_OK)
+			return M6_HIT_ERROR;
+		if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+			return M6_HIT_ERROR;
+		if (tdata1_test_rb == tdata1_hit_after_test) {
+			if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
+				return M6_HIT_ERROR;
+			return M6_NOT_HIT;
+		}
+
+		/* Check timing "imm after" support. */
+		riscv_reg_t tdata1_hit_imm_after_test = tdata1_hit_before_test | tdata1_hit_after_test;
+		if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_hit_imm_after_test) != ERROR_OK)
+			return M6_HIT_ERROR;
+		if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+			return M6_HIT_ERROR;
+		if (tdata1_test_rb == tdata1_hit_imm_after_test) {
+			if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
+				return M6_HIT_ERROR;
+			return M6_NOT_HIT;
+		}
+	}
```

Comment:
https://github.com/riscv-collab/riscv-openocd/commit/106240206e2ed388df78aed201fd61cc89ccba8f#diff-b4aa16f9e42cb8f0934baa7c8e0ec9c70a369bef98b99b26ae2e896c8aa95d6aR1587
All checks are performed under conditon `if (hit_info == CSR_MCONTROL6_HIT0_FALSE)`

Reply:
Yes. The thing is: WARL is write-any-read-legal, not write-any-read-zero. Therefore it is incorrect to assume a WARL register field is zero after a write of an illegal value to it.

In this particular case:
1. `tdata1.hit` was 0
2. Write of illegal value to `tdata1.hit`
3. Here `tdata1.hit` is any legal value which may be zero but does not have to be zero.

Reply:
We didn't assume a WARL register field is zero after a write of an illegal value to it. `tdata1_test_rb` is matched against `tdata1_...test` and `tdata1_...test` is not zero.

Reply:
I may get your point, we need to remove the check `if (tdata1_test_rb == tdata1_*_test)` in order to recover `tdata1.hit `to 0.
Thansk for catching it.

Reply:
@en-sc, my understanding is that the algorithm should be:

- All three combinations `0b01`, `0b10` and `0b11` are written to `tdata1.hit[1..0]` in a loop.
- If **any** of them results in `hit[1..0] != 0b00`, the result is `M6_NOT_HIT`.
- Otherwise the result is `M6_HIT_NOT_SUPPORTED`.

Is that please the same thing that you meant by [your comment](https://github.com/riscv-collab/riscv-openocd/pull/1144/files#r1824040455) above? Do we see it the same?

Reply:
@JanMatCodasip,

> Is that please the same thing that you meant by your comment above? Do we see it the same?


The algorithm you are describing is missing a step.
`hit` field should be set to zero after testing is done.

This is what my comment was about. Previously in the case `M6_HIT_NOT_SUPPORTED` it was not done. And looking at it once again I think **I was wrong**.
My motivation for adding the reset of `mcontrol.hit` to zero before returning `M6_HIT_NOT_SUPPORTED` was:
* The last written value to the field was illegal.
* Therefore it may contain any legal value and should be zeroed-out.

However, at this point it is known that the only legal value is zero, so `mcontrol6.hit` is zero.

@JoshCafe, @sunnyzhu-learning,  I'm sorry for the confusion. Please, consider moving
```
		if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
				return M6_HIT_ERROR;
```
into the block protected by the `if`:
```
			if (tdata1_test_rb == tdata1_tests[i]) {
				if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
					return M6_HIT_ERROR;
				return M6_NOT_HIT;
			}
```
(the same as it was before)

Reply:
- M6_NOT_HIT is returned if any of `0b01`, `0b10 `, `0b11` can be readback after writing `tdata1.hit[1..0]`.
- M6_HIT_NOT_SUPPORTED is returned if none of `0b01`, `0b10 `, `0b11` can be readback after writing `tdata1.hit[1..0]`, then the only leagal value for `hit[1..0]` is zero.
@en-sc  Thanks for your detailed explanation.

Reply:
The current version of the code looks all right to me. Thanks all.